File: src/main/python/util/Monitor.py
#
#   Copyright 2008-2015 Semantic Discovery, Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
#
import json
from datetime import datetime
from threading import Lock
from RollingStats import RollingStats

class Monitor:
    '''
    A monitor tracks processing and/or access times and rates for some function.
    '''

    def __init__(self, description=None, accessTimes=None, processingTimes=None, defaultWindowWidth=300000, defaultSegmentWidth=5000):
        '''
        NOTE: accessTimes, processingTimes should be RollingStats instances.
        '''
        self._modlock = Lock()
        self._aliveSince = datetime.now()
        self._description = description
        self._lastStartTime = None
        self._accessTimes = accessTimes
        self._processingTimes = processingTimes

        # Defaults for when creating RollingStats from within this Monitor
        self._defaultWindowWidth = defaultWindowWidth
        self._defaultSegmentWidth = defaultSegmentWidth


    @property
    def aliveSince(self):
        return self._aliveSince

    @property
    def description(self):
        return self._description

    @description.setter
    def description(self, val):
        self._description = val

    @property
    def lastAccessTime(self):
        '''
        Get the time at which access was last recorded.
        '''
        return self._lastStartTime

    @property
    def accessTimes(self):
        '''
        Get the accessTimes (RollingStats).
        '''
        return self._accessTimes

    @property
    def processingTimes(self):
        '''
        Get the processingTimes (RollingStats).
        '''
        return self._processingTimes

    @property
    def accessCumulativeStats(self):
        return None if self._accessTimes is None else self._accessTimes.cumulativeStats

    @property
    def accessWindowStats(self):
        return None if self._accessTimes is None else self._accessTimes.windowStats

    @property
    def accessWindowWidth(self):
        return None if self._accessTimes is None else self._accessTimes.windowWidth

    @property
    def processingCumulativeStats(self):
        return None if self._processingTimes is None else self._processingTimes.cumulativeStats

    @property
    def processingWindowStats(self):
        return None if self._processingTimes is None else self._processingTimes.windowStats

    @property
    def processingWindowWidth(self):
        return None if self._processingTimes is None else self._processingTimes.windowWidth

    @property
    def defaultWindowWidth(self):
        return self._defaultWindowWidth

    @defaultWindowWidth.setter
    def defaultWindowWidth(self, val):
        self._defaultWindowWidth = val

    @property
    def defaultSegmentWidth(self):
        return self._defaultSegmentWidth

    @defaultSegmentWidth.setter
    def defaultSegmentWidth(self, val):
        self._defaultSegmentWidth = val


    def summaryInfo(self):
        '''
        Get a dictionary containing a summary of this instance's information.
        '''
        result = {}

        result['aliveSince'] = str(self.aliveSince)
        if self.description is not None:
            result['description'] = self.description
        if self.lastAccessTime is not None:
            result['lastMark'] = str(self.lastAccessTime)
        if self.accessTimes is not None:
            result['accessStats'] = self.accessTimes.summaryInfo()
        if self.processingTimes is not None:
            result['processingStats'] = self.processingTimes.summaryInfo()

        return result

    def __str__(self):
        return json.dumps(self.summaryInfo(), sort_keys=True)

    def getStats(self, access=False, window=False):
        '''
        Get window/cumulative access/processing stats.
        '''
        result = None

        rollingStats = self._accessTimes if access else self._processingTimes
        if rollingStats is not None:
            result = rollingStats.windowStats if window else rollingStats.cumulativeStats

        return result

    def mark(self, starttime, endtime=None):
        '''
        Mark another access time, and processing time if endTime is not None.
        '''
        self._modlock.acquire()
        try:
            if self._lastStartTime is not None:
                if self._accessTimes is None:
                    # initialize if needed
                    self._accessTimes = RollingStats(self.defaultWindowWidth, self.defaultSegmentWidth)

                self._accessTimes.add(self._getMillis(starttime, self._lastStartTime))

            if endtime is not None:
                if self._processingTimes is None:
                    # initialize if needed
                    self._processingTimes = RollingStats(self.defaultWindowWidth, self.defaultSegmentWidth)

                self._processingTimes.add(self._getMillis(endtime, starttime))
                    
            self._lastStartTime = starttime

        finally:
            self._modlock.release()

    def _getMillis(self, laterdatetime, earlierdatetime):
        return int((laterdatetime - earlierdatetime).microseconds / 1000)

File: src/main/java/org/sd/util/Monitor.java
/*
   Copyright 2008-2016 Semantic Discovery, Inc.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/
package org.sd.util;


/**
 * Class to track processing and/or access times and rates for some function.
 * <p>
 * @author Spencer Koehler
 */
public class Monitor {
  
  private long aliveSince;
  private String description;
  private long lastAccessTime;
  private RollingStats accessTimes;
  private RollingStats processingTimes;
  private int defaultWindowWidth;
  private int defaultSegmentWidth;

  private final Object markMutex = new Object();

  public Monitor(String description) {
    this.aliveSince = System.currentTimeMillis();
    this.description = description;
    this.lastAccessTime = 0L;
    this.accessTimes = null;
    this.processingTimes = null;
    this.defaultWindowWidth = RollingStats.DEFAULT_WINDOW_WIDTH;
    this.defaultSegmentWidth = RollingStats.DEFAULT_SEGMENT_WIDTH;
  }

  public void setAliveSince(long aliveSince) {
    this.aliveSince = aliveSince;
  }

  public long getAliveSince() {
    return this.aliveSince;
  }
  
  public void setDescription(String description) {
    this.description = description;
  }

  public String getDescription() {
    return this.description;
  }
  
  public void setLastAccessTime(long lastAccessTime) {
    this.lastAccessTime = lastAccessTime;
  }

  public long getLastAccessTime() {
    return this.lastAccessTime;
  }
  
  public void setAccessTimes(RollingStats accessTimes) {
    this.accessTimes = accessTimes;
  }

  public RollingStats getAccessTimes() {
    return this.accessTimes;
  }
  
  public void setProcessingTimes(RollingStats processingTimes) {
    this.processingTimes = processingTimes;
  }

  public RollingStats getProcessingTimes() {
    return this.processingTimes;
  }
  
  public void setDefaultWindowWidth(int defaultWindowWidth) {
    this.defaultWindowWidth = defaultWindowWidth;
  }

  public int getDefaultWindowWidth() {
    return this.defaultWindowWidth;
  }
  
  public void setDefaultSegmentWidth(int defaultSegmentWidth) {
    this.defaultSegmentWidth = defaultSegmentWidth;
  }

  public int getDefaultSegmentWidth() {
    return this.defaultSegmentWidth;
  }
  

  public void mark(long starttime) {
    this.mark(starttime, 0L);
  }

  public void mark(long starttime, long endtime) {
    synchronized (this.markMutex) {
      if (this.lastAccessTime > 0) {
        if (this.accessTimes == null) {
          this.accessTimes = new RollingStats(this.defaultWindowWidth, this.defaultSegmentWidth);
        }
        this.accessTimes.add(starttime - this.lastAccessTime);
      }

      if (endtime > 0) {
        if (this.processingTimes == null) {
          this.processingTimes = new RollingStats(this.defaultWindowWidth, this.defaultSegmentWidth);
        }
        this.processingTimes.add(endtime - starttime);
      }

      this.lastAccessTime = starttime;
    }
  }

  public StatsAccumulator getStats(boolean access, boolean window) {
    // Get window/cumulative access/processing stats
    StatsAccumulator result = null;

    final RollingStats rollingStats = access ? this.accessTimes : this.processingTimes;
    if (rollingStats != null) {
      result = window ? rollingStats.getWindowStats() : rollingStats.getCumulativeStats();
    }

    return result;
  }

  public StatsAccumulator getAccessWindowStats() {
    return getStats(true, true);
  }

  public StatsAccumulator getAccessCumulativeStats() {
    return getStats(true, false);
  }

  public StatsAccumulator getProcessingWindowStats() {
    return getStats(false, true);
  }

  public StatsAccumulator getProcessingCumulativeStats() {
    return getStats(false, false);
  }
}
